File: src/androidTest/java/com/astoev/cave/survey/test/service/data/ExcelImportTest.java
package com.astoev.cave.survey.test.service.data;

import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.filters.LargeTest;

import com.astoev.cave.survey.service.imp.ExcelImport;
import com.astoev.cave.survey.service.imp.LegData;
import com.astoev.cave.survey.service.imp.ProjectData;

import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.IOException;
import java.io.InputStream;
import java.util.List;

import static com.astoev.cave.survey.model.Option.UNIT_GRADS;
import static com.astoev.cave.survey.model.Option.UNIT_METERS;
import static com.astoev.cave.survey.test.helper.ExcelTestUtils.assertConfigUnits;
import static com.astoev.cave.survey.test.helper.ExcelTestUtils.assertDefaultUnits;
import static com.astoev.cave.survey.test.helper.ExcelTestUtils.assertLeg;
import static com.astoev.cave.survey.test.helper.ExcelTestUtils.assertLegLocation;
import static org.junit.Assert.assertEquals;

@RunWith(AndroidJUnit4.class)
@LargeTest
public class ExcelImportTest {

    @Test
    public void testSimpleFile() throws IOException {

        // file properly loaded
        ProjectData data = loadExcel("simple_20190220_1");

        // proper units
        assertConfigUnits(data, UNIT_METERS, UNIT_GRADS, UNIT_GRADS);

        // check some data
        List<LegData> legs = data.getLegs();
        assertEquals(1, legs.size());
        assertLeg(legs.get(0), 1f, 2f, 3f, 1f, 3f, 4f, 2f);
        assertLegLocation(legs.get(0), null, null, null, null);
    }

    @Test
    public void testImportExportedFile() throws IOException {

        // file properly loaded
        ProjectData data = loadExcel("Stefan_prekop_20190207_1_initial");

        // proper units
        assertDefaultUnits(data);

        // check some data
        List<LegData> legs = data.getLegs();
        assertEquals(105, legs.size());
        assertLeg(legs.get(0), 6f, 231.1f, 0.626f);
        assertLegLocation(legs.get(0), 48.324408f, 17.224757f, 399f, 10f);
        assertLeg(legs.get(1), 11f, 236.4f, 0.98f);
        assertLegLocation(legs.get(1), null, null, null, null);
    }

    @Test
    public void testImportEditedFileWithStringCellValues() throws IOException {

        // file properly loaded
        ProjectData data = loadExcel("Stefan_prekop_20190207_1_stringcells");

        // proper units
        assertDefaultUnits(data);

        // check some data
        List<LegData> legs = data.getLegs();
        assertEquals(106, legs.size());
        assertLeg(legs.get(0), 6f, 231.1f, 0f);
        assertLegLocation(legs.get(0), 48.324408f, 17.224757f, 399f, 10f);
        assertLeg(legs.get(1), 11f, 236.4f, 0f);
        assertLegLocation(legs.get(1), null, null, null, null);
    }

    private ProjectData loadExcel(String aFile) throws IOException {
        InputStream resourceStream = this.getClass().getClassLoader().getResourceAsStream("/export/xls/" + aFile + ".xls");
        ProjectData data = ExcelImport.loadProjectData(resourceStream);
        return data;
    }
}

File: src/androidTest/java/com/astoev/cave/survey/test/service/data/ExcelImportTest.kt
package com.astoev.cave.survey.test.service.data

import com.astoev.cave.survey.model.Option.*
import com.astoev.cave.survey.service.imp.ExcelImport
import com.astoev.cave.survey.service.imp.ProjectData
import com.astoev.cave.survey.test.helper.ExcelTestUtils
import org.junit.Assert
import org.junit.Test
import java.io.IOException

class ExcelImportTest : AbstractExportTest() {

    @Test
    @Throws(IOException::class)
    fun testSimpleFile() {

        // file properly loaded
        val data = loadExcel("simple_20190220_1")

        // proper units
        ExcelTestUtils.assertConfigUnits(data, UNIT_METERS, UNIT_DEGREES, UNIT_DEGREES)

        // check some data
        val legs = data.legs
        Assert.assertEquals(1, legs.size.toLong())
        ExcelTestUtils.assertLeg(legs[0], 1f, 2f, 3f, 1f, 3f, 4f, 2f)
        ExcelTestUtils.assertLegLocation(legs[0], null, null, null, null)
    }

    @Test
    @Throws(IOException::class)
    fun testSimpleFileNonDefaultUnits() {

        // file properly loaded
        val data = loadExcel("simple_20190220_non_default_units")

        // proper units
        ExcelTestUtils.assertConfigUnits(data, UNIT_FEET, UNIT_GRADS, UNIT_GRADS)

        // check some data
        val legs = data.legs
        Assert.assertEquals(1, legs.size.toLong())
        ExcelTestUtils.assertLeg(legs[0], 1f, 2f, 3f, 1f, 3f, 4f, 2f)
        ExcelTestUtils.assertLegLocation(legs[0], null, null, null, null)
    }

    @Test
    @Throws(IOException::class)
    fun testImportExportedFile() {

        // file properly loaded
        val data = loadExcel("Stefan_prekop_20190207_1_initial")

        // proper units
        ExcelTestUtils.assertDefaultUnits(data)

        // check some data
        val legs = data.legs
        Assert.assertEquals(105, legs.size.toLong())
        ExcelTestUtils.assertLeg(legs[0], 6f, 231.1f, 0.626f)
        ExcelTestUtils.assertLegLocation(legs[0], 48.324408, 17.224757, 399.0, 10.0)
        ExcelTestUtils.assertLeg(legs[1], 11f, 236.4f, 0.98f)
        ExcelTestUtils.assertLegLocation(legs[1], null, null, null, null)
    }

    @Test
    @Throws(IOException::class)
    fun testImportEditedFileWithStringCellValues() {

        // file properly loaded
        val data = loadExcel("Stefan_prekop_20190207_1_stringcells")

        // proper units
        ExcelTestUtils.assertDefaultUnits(data)

        // check some data
        val legs = data.legs
        Assert.assertEquals(106, legs.size.toLong())
        ExcelTestUtils.assertLeg(legs[0], 6f, 231.1f, 0f)
        ExcelTestUtils.assertLegLocation(legs[0], 48.324408, 17.224757, 399.0, 10.0)
        ExcelTestUtils.assertLeg(legs[1], 11f, 236.4f, 0f)
        ExcelTestUtils.assertLegLocation(legs[1], null, null, null, null)
    }

    @Throws(IOException::class)
    private fun loadExcel(aFile: String): ProjectData {
        val resourceStream = findAsset("export/xls/$aFile.xls")
        return ExcelImport.loadProjectData(resourceStream)
    }

    @Test
    @Throws(IOException::class)
    fun testImportEmptyRowsAndNumericCoordinates() {

        // file properly loaded
        val data = loadExcel("numeric_coordinate_cells_and_empty_rows")

        // proper units
        ExcelTestUtils.assertConfigUnits(data, UNIT_METERS, UNIT_GRADS, UNIT_GRADS)
        // check some data
        val legs = data.legs
        Assert.assertEquals(3, legs.size.toLong())
        ExcelTestUtils.assertLeg(legs[0], 4.73f, 235f, -3f, 5f, 1f, 4f, 6f)
        ExcelTestUtils.assertLegLocation(legs[0], 41.945457, 24.022383, null, null)
        ExcelTestUtils.assertLeg(legs[1], 5.22f, 135f, -85f)
        ExcelTestUtils.assertLegLocation(legs[1], null, null, null, null)
        ExcelTestUtils.assertLeg(legs[2], 3.1f, 0f, -90f)
        ExcelTestUtils.assertLegLocation(legs[2], null, null, null, null)
    }
}